File: oss/clawpipe/ruby-sdk/lib/clawpipe/packer.rb
# frozen_string_literal: true

module ClawPipe
  PackResult = Struct.new(:packed, :original_tokens, :packed_tokens, :savings, keyword_init: true)

  PackerConfig = Struct.new(
    :max_tokens, :deduplication, :strip_boilerplate, :compress_whitespace,
    keyword_init: true
  ) do
    def self.default
      new(max_tokens: 100_000, deduplication: true, strip_boilerplate: true, compress_whitespace: true)
    end
  end

  # Compresses context to reduce token count.
  # Strategies: whitespace compression, deduplication, boilerplate stripping, budget truncation.
  class Packer
    def initialize(config = nil)
      @config = config || PackerConfig.default
    end

    def pack(input_text, system: nil)
      original = system ? "#{system}\n\n#{input_text}" : input_text
      original_tokens = estimate_tokens(original)

      packed = original
      packed = compress_whitespace(packed) if @config.compress_whitespace
      packed = deduplicate(packed) if @config.deduplication
      packed = strip_boilerplate(packed) if @config.strip_boilerplate
      packed = truncate_to_limit(packed)

      packed_tokens = estimate_tokens(packed)
      savings_pct = original_tokens > 0 ? ((1.0 - packed_tokens.to_f / original_tokens) * 100).round : 0

      PackResult.new(
        packed: packed,
        original_tokens: original_tokens,
        packed_tokens: packed_tokens,
        savings: "#{[0, savings_pct].max}%"
      )
    end

    def estimate_tokens(text)
      (text.length / 4.0).ceil
    end

    private

    def compress_whitespace(text)
      lines = text.split("\n").map(&:rstrip)
      joined = lines.join("\n")
      joined.gsub(/\n{3,}/, "\n\n").strip
    end

    def deduplicate(text)
      blocks = text.split("\n\n")
      seen = {}
      unique = []
      blocks.each do |block|
        normalized = block.strip.downcase
        next if normalized.empty?
        next if normalized.length > 50 && seen[normalized]

        seen[normalized] = true
        unique << block
      end
      unique.join("\n\n")
    end

    BOILERPLATE_PATTERNS = [
      /^\/\*\*?\s*\n(\s*\*\s*@(param|returns|throws|example).*\n)*\s*\*\//,
      /^\/\/\s*eslint-disable.*$/,
      /^\/\/\s*@ts-(ignore|expect-error|nocheck).*$/,
      /^'use strict';?\s*$/,
      /^\/\*\s*istanbul ignore (next|else)\s*\*\/$/,
    ].freeze

    def strip_boilerplate(text)
      result = text
      BOILERPLATE_PATTERNS.each { |pat| result = result.gsub(pat, '') }
      result.gsub(/\n{3,}/, "\n\n").strip
    end

    def truncate_to_limit(text)
      max_chars = @config.max_tokens * 4
      return text if text.length <= max_chars

      truncated = text[0, max_chars]
      last_nl = truncated.rindex("\n")
      cut = (last_nl && last_nl > max_chars * 0.8) ? last_nl : max_chars
      "#{truncated[0, cut]}\n\n[Truncated -- context exceeded budget]"
    end
  end
end

File: oss/clawpipe/python-sdk/clawpipe/packer.py
"""Context Packer -- compress context to reduce token count.

Strategies: whitespace compression, deduplication, boilerplate
stripping, and budget truncation.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass


@dataclass
class PackResult:
    """Result of packing a prompt."""

    packed: str
    original_tokens: int
    packed_tokens: int
    savings: str


@dataclass
class PackerConfig:
    """Packer configuration."""

    max_tokens: int = 100_000
    deduplication: bool = True
    strip_boilerplate: bool = True
    compress_whitespace: bool = True


class Packer:
    """Compress context to reduce token count."""

    def __init__(self, config: PackerConfig | None = None) -> None:
        self._config = config or PackerConfig()

    def pack(self, input_text: str, system: str | None = None) -> PackResult:
        """Pack a prompt and optional system message."""
        original = f"{system}\n\n{input_text}" if system else input_text
        original_tokens = self.estimate_tokens(original)

        packed = original

        if self._config.compress_whitespace:
            packed = self._compress_whitespace(packed)

        if self._config.deduplication:
            packed = self._deduplicate(packed)

        if self._config.strip_boilerplate:
            packed = self._strip_boilerplate(packed)

        packed = self._truncate_to_limit(packed)

        packed_tokens = self.estimate_tokens(packed)
        savings_pct = (
            round((1 - packed_tokens / original_tokens) * 100)
            if original_tokens > 0
            else 0
        )

        return PackResult(
            packed=packed,
            original_tokens=original_tokens,
            packed_tokens=packed_tokens,
            savings=f"{max(0, savings_pct)}%",
        )

    @staticmethod
    def estimate_tokens(text: str) -> int:
        """Rough token estimate: ~4 chars per token."""
        return math.ceil(len(text) / 4)

    def _compress_whitespace(self, text: str) -> str:
        lines = [line.rstrip() for line in text.split("\n")]
        joined = "\n".join(lines)
        return re.sub(r"\n{3,}", "\n\n", joined).strip()

    def _deduplicate(self, text: str) -> str:
        blocks = text.split("\n\n")
        seen: set[str] = set()
        unique: list[str] = []
        for block in blocks:
            normalized = block.strip().lower()
            if not normalized:
                continue
            if len(normalized) > 50 and normalized in seen:
                continue
            seen.add(normalized)
            unique.append(block)
        return "\n\n".join(unique)

    def _strip_boilerplate(self, text: str) -> str:
        patterns = [
            re.compile(
                r"^/\*\*?\s*\n(\s*\*\s*@(param|returns|throws|example).*\n)*\s*\*/",
                re.MULTILINE,
            ),
            re.compile(r"^//\s*eslint-disable.*$", re.MULTILINE),
            re.compile(r"^//\s*@ts-(ignore|expect-error|nocheck).*$", re.MULTILINE),
            re.compile(r"^'use strict';?\s*$", re.MULTILINE),
            re.compile(r"^/\*\s*istanbul ignore (next|else)\s*\*/$", re.MULTILINE),
        ]
        result = text
        for pat in patterns:
            result = pat.sub("", result)
        return re.sub(r"\n{3,}", "\n\n", result).strip()

    def _truncate_to_limit(self, text: str) -> str:
        max_chars = self._config.max_tokens * 4
        if len(text) <= max_chars:
            return text
        truncated = text[:max_chars]
        last_nl = truncated.rfind("\n")
        cut = last_nl if last_nl > max_chars * 0.8 else max_chars
        return truncated[:cut] + "\n\n[Truncated -- context exceeded budget]"
